pytorch_pfn_extras/training/extensions/slack.py

```diff
@@ -0,0 +1,205 @@
+import os
+import json
+import warnings
+
+from typing import Any, Callable, List, Optional, Union
+
+from pytorch_pfn_extras.training import extension
+from pytorch_pfn_extras.training import trigger as trigger_module
+from pytorch_pfn_extras.training._manager_protocol import ExtensionsManagerProtocol
+
+try:
+    import requests  # type: ignore[import]
+    _requests_available = True
+except ImportError:
+    _requests_available = False
+
+try:
+    import slack_sdk
+    _slack_sdk_available = True
+except ImportError:
+    _slack_sdk_available = False
+
+
+class Slack(extension.Extension):
+    """An extension to communicate with Slack.
+
+    This extension receives a ``text_template`` argument that contains
+    placeholders that will be populated with ``manager`` or ``manager.observation``
+    fields, and custom objects such as ``context_object``.
+
+    For example. `text_template = "Loss {loss} for iteration {.iteration}"`
+    will be populated as `text_template.format(manager, context_object, **observation)`
+    retrieving the ``loss`` value from the ``observation`` and ``.iteration`` from
+    the manager object. Instead of string, a callable object taking the
+    ``ExtensionsManager``, ``context_object`` and the observation dictionary can
+    be used instead. The same rules apply for the ``filenames_template`` argument.
+
+    Args:
+        channel_id (str): The channel where messages or files will be sent.
+        text_template (str or callable): Template for sending the message.
+            It can be a string to be formatted using ``.format`` or a callable
+            that returns a string. In both cases, `manager`, the current
+            observation dict and the context object are used to look up
+            values of interest.
+        start_text_template (str or callable): Template for sending a message
+            at the beggining of the experiment.
+            See ``text_template`` for format.
+        end_text_template (str or callable): Template for sending a message
+            at the completion of the experiment.
+            See ``text_template`` for format.
+        filenames_template (list of str or callable): list of files that will
+            be uploaded to slack, these are string templates that can take
+            values in the same way as ``text_template``. Optional.
+        use_threads (bool): If subsequent calls of this extensions should be
+            posted as a thread of the original message or not.
+            Default is ``False``.
+        context_object (object): Custom object that contains data used to
+            format the text or the filenames. Optional, default is ``None``.
+        webhook_url (str): Webhook to send messages, it is mutually exclusive
+            with ``token`` and ``client``. Used when slack_sdk is not available.
```

Comment:
The purpose we provide webhook pai is to allow quickly setting up intergation.

```suggestion
            with ``token`` and ``client``.
```